pkg/client/attach.go

```diff
@@ -0,0 +1,257 @@
+			return errors.Wrapf(err, "failed to connect to container's attach socket: %v", cfg.SocketPath)
+		}
+		defer func() {
+			if err := conn.Close(); err != nil {
+				logrus.Errorf("unable to close socket: %q", err)
+			}
+		}()
+	}
+
+	if cfg.StartFunc != nil {
+		if err := cfg.StartFunc(); err != nil {
+			return err
+		}
+	}
+
+	if cfg.Passthrough {
+		return nil
+	}
+
+	receiveStdoutError, stdinDone := setupStdioChannels(cfg, conn)
+	if cfg.AttachReady != nil {
+		cfg.AttachReady <- true
+	}
+	return readStdio(cfg, conn, receiveStdoutError, stdinDone)
+}
+func setupStdioChannels(cfg *AttachConfig, conn *net.UnixConn) (chan error, chan error) {
+	receiveStdoutError := make(chan error)
+	go func() {
+		receiveStdoutError <- redirectResponseToOutputStreams(cfg, conn)
+	}()
+
+	stdinDone := make(chan error)
+	go func() {
+		var err error
+		if cfg.Streams.AttachStdin {
+			_, err = utils.CopyDetachable(conn, cfg.Streams.Stdin, cfg.DetachKeys)
+		}
+		stdinDone <- err
+	}()
+
+	return receiveStdoutError, stdinDone
+}
+
+func redirectResponseToOutputStreams(cfg *AttachConfig, conn io.Reader) error {
+	var err error
+	buf := make([]byte, 8192+1) /* Sync with conmon STDIO_BUF_SIZE */
+	for {
+		nr, er := conn.Read(buf)
+		if nr > 0 {
+			var dst io.Writer
+			var doWrite bool
+			switch buf[0] {
+			case AttachPipeStdout:
+				dst = cfg.Streams.Stdout
+				doWrite = cfg.Streams.AttachStdout
+			case AttachPipeStderr:
+				dst = cfg.Streams.Stderr
+				doWrite = cfg.Streams.AttachStderr
+			default:
+				logrus.Infof("Received unexpected attach type %+d", buf[0])
+			}
+			if dst == nil {
+				return errors.New("output destination cannot be nil")
+			}
+
+			if doWrite {
+				nw, ew := dst.Write(buf[1:nr])
+				if ew != nil {
+					err = ew
+					break
+				}
+				if nr != nw+1 {
+					err = io.ErrShortWrite
+					break
+				}
+			}
+		}
+		if er == io.EOF {
+			break
+		}
+		if er != nil {
+			err = er
+			break
+		}
+	}
+	return err
+}
+
+func readStdio(cfg *AttachConfig, conn *net.UnixConn, receiveStdoutError, stdinDone chan error) error {
+	var err error
+	select {
+	case err = <-receiveStdoutError:
+		conn.CloseWrite()
+		return err
+	case err = <-stdinDone:
+		// This particular case is for when we get a non-tty attach
+		// with --leave-stdin-open=true. We want to return as soon
+		// as we receive EOF from the client. However, we should do
+		// this only when stdin is enabled. If there is no stdin
+		// enabled then we wait for output as usual.
+		if cfg.DetachStdin {
+			return nil
+		}
+		if err == define.ErrDetach {
+			conn.CloseWrite()
+			return err
+		}
+		if err == nil {
+			// copy stdin is done, close it
+			if connErr := conn.CloseWrite(); connErr != nil {
+				logrus.Errorf("Unable to close conn: %v", connErr)
+			}
+		}
+		if cfg.Streams.AttachStdout || cfg.Streams.AttachStderr {
+			return <-receiveStdoutError
+		}
+	}
+	return nil
+}
+
+type SetWindowSizeContainerConfig struct {
+	ID     string
+	Width  uint16
+	Height uint16
```

Comment:
Let's add comments here too, it's the height/width in rows/columns.

Reply:
I just dropped the struct